### PR TITLE
Downstream add two factor exception

### PR DESCRIPTION
**Pull request by LukasReschke**
From https://github.com/owncloud/core/pull/26628

### CODE REVIEW

File: core/Controller/TwoFactorChallengeController.php [hunk 1]
```diff
@@ -115,16 +116,20 @@ public function showChallenge($challengeProviderId, $redirect_url) {
 			$backupProvider = null;
 		}
 
+		$error_message = "";
```
Comment (LukasReschke) [inline core/Controller/TwoFactorChallengeController.php]: Single quotes

Comment (LukasReschke) [inline core/Controller/TwoFactorChallengeController.php]: Also change to `$errorMessage`

File: core/Controller/TwoFactorChallengeController.php [hunk 2]
```diff
@@ -150,11 +155,21 @@ public function solveChallenge($challengeProviderId, $challenge, $redirect_url =
...
 			}
-			return new RedirectResponse(OC_Util::getDefaultPageUrl());
+		} catch (TwoFactorException $e) {
+			/*
+			 * The 2FA App threw an TwoFactorException. Now we display more
+			 * information to the user. The exception text is stored in the
+			 * session to be used in showChallenge()
+			 */
+			$this->session->set('two_factor_auth_error_message',
+				$e->getMessage());
```
Comment (LukasReschke) [inline core/Controller/TwoFactorChallengeController.php]: Move to one line

File: lib/public/Authentication/TwoFactorAuth/TwoFactorException.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ * User: cornelius
+ * Date: 14.11.16
```
Comment (LukasReschke) [inline lib/public/Authentication/TwoFactorAuth/TwoFactorException.php]: Change to default oC AGPL header.

File: lib/public/Authentication/TwoFactorAuth/TwoFactorException.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * display this error.
+ */
+
+// use OCP namespace for all classes that are considered public.
+// This means that they should be used by apps instead of the internal ownCloud classes
+namespace OCP\Authentication\TwoFactorAuth;
+
+/**
+ * Two Factor Authentication failed
+ * @since 9.2.0
```
Comment (LukasReschke) [inline lib/public/Authentication/TwoFactorAuth/TwoFactorException.php]: 11.0

Comment (rullzer) [inline lib/public/Authentication/TwoFactorAuth/TwoFactorException.php]: 12.0

Comment (LukasReschke) [inline lib/public/Authentication/TwoFactorAuth/TwoFactorException.php]: True.

File: core/Controller/TwoFactorChallengeController.php [hunk 5]
```diff
@@ -150,11 +155,21 @@ public function solveChallenge($challengeProviderId, $challenge, $redirect_url =
 			return new RedirectResponse($this->urlGenerator->linkToRoute('core.TwoFactorChallenge.selectChallenge'));
 		}
 
-		if ($this->twoFactorManager->verifyChallenge($challengeProviderId, $user, $challenge)) {
-			if (!is_null($redirect_url)) {
-				return new RedirectResponse($this->urlGenerator->getAbsoluteURL(urldecode($redirect_url)));
+		try {
+			if ($this->twoFactorManager->verifyChallenge($challengeProviderId, $user, $challenge)) {
```
Comment (LukasReschke) [inline core/Controller/TwoFactorChallengeController.php]: Codpath is not covered by tests.

Comment (ChristophWurst) [inline core/Controller/TwoFactorChallengeController.php]: done

File: core/templates/twofactorshowchallenge.php [hunk 6]
```diff
@@ -12,7 +14,11 @@
 <div class="warning">
 		<h2 class="two-factor-header"><?php p($provider->getDisplayName()); ?></h2>
 		<?php if ($error): ?>
-		<p><strong><?php p($l->t('Error while validating your second factor')); ?></strong></p>
+			<?php if($error_message): ?>
```
Comment (ChristophWurst) [inline core/templates/twofactorshowchallenge.php]: @LukasReschke what are we gonna do in terms of l10n? I presume the exception's message won't be translated.

Comment (rullzer) [inline core/templates/twofactorshowchallenge.php]: Stuff like that should be localized.

What comes to mind:

* Have TwoFactorException extend the HintException.
* Make sure that apps when throwing exceptions set the hint localized
* Show the Hint

Comment (LukasReschke) [inline core/templates/twofactorshowchallenge.php]: IMO the error message should be localized itself, so that should not be a problem.

Comment (rullzer) [inline core/templates/twofactorshowchallenge.php]: Also fine. Just localize it ;)

**Comment by ChristophWurst** [comment]
conflicts resolved and rebased

**Comment by rullzer** [comment]
Updated the autoloader.

**Review by rullzer — APPROVED** [review]
I'd say get this in then we can move forward.

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**